File: Sorts/Quick Sort/Quick Sort.cpp
#include <iostream>

using namespace std;
///
// Start sort function
///
void QuickSort(int arr[], int left, int right) {
	int i = left, j = right;
	int middle = arr[(left + right) / 2];
	while (i <= j) {
		while (arr[i] < middle)
			i++;
		while (arr[j] > middle)
			j--;
		if (i <= j)
			swap(arr[i++], arr[j--]);
	}
	if (left < j)
		QuickSort(arr, left, j);
	if (i < right)
		QuickSort(arr, i, right);
}
///
// End of sort function
///
int main()
{
	///
	// Exemple of usage
	///
	int *arr;
	int size;
	cout << "n = ";
	cin >> size;

	arr = new int[size];


	///
	// Read array
	///
	for (int i = 0; i < size; i++) {
		cout << "arr[" << i << "] = ";
		cin >> arr[i];
	}
	///
	// Calling a sort function
	///
	QuickSort(arr,-1, size-1);


	///
	// Output result
	///
	for (int i = 0; i < size; i++) {
		cout << arr[i] << " ";
	}
	cout << endl;
	delete[] arr;
	return 0;
	///
	//  End of exemple of usage
	///
}

File: Sorts/Quick Sort/Quick Sort.py
###
## Start sort function
###
def QuickSort(arr, left, right):
    i = left
    j = right
    middle = arr[(left + right) // 2]
    while i <= j:
        while arr[i] < middle:
            i += 1
        while arr[j] > middle:
            j -= 1
        if i <= j:
            arr[i], arr[j] = arr[j], arr[i]
            i += 1
            j -= 1
    if left < j:
        QuickSort(arr, left, j)
    if i < right:
        QuickSort(arr, i, right)
###
## End of sort function
###



###
## Exemple of usage
###

arr = []
size = int(input("n = "))

###
## Read array
###

i = 0
while i < size:
    arr.append(int(input("arr[" + str(i) + "] = ")))
    i += 1

###
## Calling a sort function
###

QuickSort(arr,0, size - 1)

###
## Output result
###

print(arr)

###
##  End of exemple of usage
###
